pkg/parse/events/sources.go

```diff
@@ -0,0 +1,232 @@
+func (s *TimeDelaySource) Start(ctx context.Context) reflect.Value {
+	s.timer = s.Clock.NewTimer(s.Period)
+	go func() {
+		<-ctx.Done()
+		s.timer.Stop()
+	}()
+	return reflect.ValueOf(s.timer.C())
+}
+
+// HandleEvent calls the HandleFunc with a new event and resets the delay timer.
+func (s *TimeDelaySource) HandleEvent(handleFn HandleFunc) Result {
+	result := handleFn(Event{Type: s.EventType})
+
+	// Schedule next resync attempt
+	s.timer.Reset(s.Period)
+	return result
+}
+
+// HandleResult is a no-op.
+func (s *TimeDelaySource) HandleResult(_ Result) {}
+
+// NewRetrySyncSource constructs an TimeDelaySourceWithBackoff that generates
+// and handles RetrySyncEvents with retry backoff.
+func NewRetrySyncSource(c clock.Clock, backoff wait.Backoff) Source {
+	return &TimeDelaySourceWithBackoff{
+		EventType: RetrySyncEventType,
+		Clock:     c,
+		Backoff:   backoff,
+	}
+}
+
+// TimeDelaySourceWithBackoff sends StatusEvents periodically using a backoff
+// timer that is incremented after each event is handled.
+//
+// Unlike, TimeDelaySource, TimeDelaySourceWithBackoff always increases the
+// delay, unless the subscriber sets Result.ResetRetryBackoff.
+type TimeDelaySourceWithBackoff struct {
+	EventType EventType
+	Clock     clock.Clock
+	Backoff   wait.Backoff
+
+	currentBackoff wait.Backoff
+	retryLimit     int
+	timer          clock.Timer
+}
+
+// Type of events produced by this source.
+func (s *TimeDelaySourceWithBackoff) Type() EventType {
+	return s.EventType
+}
+
+// Start the timer and return the event channel.
+func (s *TimeDelaySourceWithBackoff) Start(ctx context.Context) reflect.Value {
+	s.currentBackoff = util.CopyBackoff(s.Backoff)
+	s.retryLimit = s.currentBackoff.Steps
+	s.timer = s.Clock.NewTimer(s.currentBackoff.Duration)
+	go func() {
+		<-ctx.Done()
+		s.timer.Stop()
+	}()
+	return reflect.ValueOf(s.timer.C())
+}
+
+// HandleEvent calls the HandleFunc with a new event, increments the backoff
+// step, and updates the delay timer.
+// If the maximum number of retries has been reached, the HandleFunc is NOT
+// called and an empty Result is returned.
+func (s *TimeDelaySourceWithBackoff) HandleEvent(handleFn HandleFunc) Result {
+	if s.currentBackoff.Steps == 0 {
+		klog.Infof("Retry limit (%v) has been reached", s.retryLimit)
+		// Don't reset retryTimer if retry limit has been reached.
+		return Result{}
+	}
+
+	retryDuration := s.currentBackoff.Step()
+	retries := s.retryLimit - s.currentBackoff.Steps
+	klog.Infof("a retry is triggered (retries: %v/%v)", retries, s.retryLimit)
+
+	result := handleFn(Event{Type: s.EventType})
+
+	// Schedule next resync attempt
+	s.timer.Reset(retryDuration)
+	return result
+}
+
+// HandleResult resets the backoff timer if ResetRetryBackoff is true.
+func (s *TimeDelaySourceWithBackoff) HandleResult(result Result) {
+	if result.ResetRetryBackoff {
+		s.currentBackoff = util.CopyBackoff(s.Backoff)
+		s.timer.Reset(s.currentBackoff.Duration)
```

Comment:
I understand if it's a lot to ask - But it might be nice to have a table of `scenarios x timers` (where each cell indicates whether the scenario causes the timer to be reset) to help visualize which scenarios/events cause which timers to get reset. Would help clarify the behavior

Reply:
I added events/doc.go with more details.